sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala

```diff
@@ -2261,6 +2261,32 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
     assert(ex3.getMessage.contains("cannot resolve 'a'"))
   }
 
+  test("nested transform (DSL)") {
+    val df = Seq(
+        (Seq(1, 2, 3), Seq("a", "b", "c"))
+    ).toDF("numbers", "letters")
+
+    checkAnswer(
+        df.select(
+          flatten(
+            transform(
+              $"numbers",
+              (number: Column) => transform(
```

Comment:
I think we need exhaustive tests for this issue, e.g., two/three argument cases and their combination cases.